File: cpptet_v0.5/Matrix.cpp
#include "Matrix.h"

void Matrix::alloc(int cy, int cx) { //2차원 배열 동적할당하는 함수
  if ((cy < 0) || (cx < 0)) return;
  dy = cy;
  dx = cx;
  array = new int*[dy]; //y축 동적할당
  for (int y = 0; y < dy; y++) //반복문 통해 x축 동적할당
    array[y] = new int[dx]; 
  for (int y = 0; y < dy; y++) //2차원 배열 모든 값 0으로 초기화
    for (int x = 0; x < dx; x++)
      array[y][x] = 0;
}

int Matrix::get_dy() { return dy; } 

int Matrix::get_dx() { return dx; }

int **Matrix::get_array() { return array; }

Matrix::Matrix() { alloc(0, 0); }

Matrix::~Matrix() { 
  for (int y = 0; y < dy; y++)
    delete array[y];
  delete array;
}

Matrix::Matrix(int cy, int cx) {
  alloc(cy, cx);
  for (int y = 0; y < dy; y++) //이건 alloc에도 있어서 안해도 될듯?
    for (int x = 0; x < dx; x++)
      array[y][x] = 0;
}

Matrix::Matrix(const Matrix *obj) { //포인터 통한 copy constructor
  alloc(obj->dy, obj->dx);
  for (int y = 0; y < dy; y++)
    for (int x = 0; x < dx; x++)
      array[y][x] = obj->array[y][x];
}

Matrix::Matrix(const Matrix &obj) { //참조자 통한 copy constructor
  alloc(obj.dy, obj.dx);
  for (int y = 0; y < dy; y++)
    for (int x = 0; x < dx; x++)
      array[y][x] = obj.array[y][x];
}

Matrix::Matrix(int *arr, int col, int row) { //1차원 배열로 저장되어있는 블록 2차원 행렬로 생성
  alloc(col, row);
  for (int y = 0; y < dy; y++)
    for (int x = 0; x < dx; x++)
      array[y][x] = arr[y * dx + x];
}

Matrix *Matrix::clip(int top, int left, int bottom, int right) { //행렬을 부분적으로 자르는 함수
  int cy = bottom - top;
  int cx = right - left;
  Matrix *temp = new Matrix(cy, cx);
  for (int y = 0; y < cy; y++) {
    for (int x = 0; x < cx; x++) {
      if ((top + y >= 0) && (left + x >= 0) &&
	  (top + y < dy) && (left + x < dx))
	temp->array[y][x] = array[top + y][left + x];
      else {
	cerr << "invalid matrix range";
	return NULL;
      }
    }
  }
  return temp; //자른 행렬 반환
}

void Matrix::paste(const Matrix *obj, int top, int left) { //다른 행렬을 현재 행렬에 붙여주는 함수
  for (int y = 0; y < obj->dy; y++)
    for (int x = 0; x < obj->dx; x++) {
      if ((top + y >= 0) && (left + x >= 0) &&
	  (top + y < dy) && (left + x < dx))
	array[y + top][x + left] = obj->array[y][x];
      else {
	cerr << "invalid matrix range";
	return NULL;
      }
    }
}

Matrix *Matrix::add(const Matrix *obj) { //현재행렬에 다른 행렬의 값을 더하는 함수
  if ((dx != obj->dx) || (dy != obj->dy)) return NULL;
  Matrix *temp = new Matrix(dy, dx);
  for (int y = 0; y < dy; y++)
    for (int x = 0; x < dx; x++)
      temp->array[y][x] = array[y][x] + obj->array[y][x];
  return temp; //더한 후의 행렬 반환
}

int Matrix::sum() { //행렬에 있는 모든 원소의 합 반환해주는 함수
  int total = 0;
  for (int y = 0; y < dy; y++)
    for (int x = 0; x < dx; x++)
      total += array[y][x];
  return total;
}

void Matrix::mulc(int coef) { //현재 행렬의 모든 원소에 coef만큼 상수배 곱해주는 함수
  for (int y = 0; y < dy; y++)
    for (int x = 0; x < dx; x++)
      array[y][x] = coef * array[y][x];
}

Matrix *Matrix::binary() { //현재 행렬을 이용해 원소가 0과 1로만 이루어진 행렬을 반환해주는 함수
  Matrix *temp = new Matrix(dy, dx);
  int **t_array = temp->get_array();
  for (int y = 0; y < dy; y++)
    for (int x = 0; x < dx; x++)
      t_array[y][x] = (array[y][x] != 0 ? 1 : 0); //0이면 0, 0이 아니면 1
  
  return temp;
}

bool Matrix::anyGreaterThan(int val) { //행렬에 val보다 큰 값이 있는지 확인하는 함수
  for (int y = 0; y < dy; y++) {
    for (int x = 0; x < dx; x++) {
      if (array[y][x] > val)
	return true; //있으면 true
    }
  }
  return false; //없으면 false를 반환
}

void Matrix::print() { //행렬 출력 함수
  cout << "Matrix(" << dy << "," << dx << ")" << endl;
  for (int y = 0; y < dy; y++) {
    for (int x = 0; x < dx; x++)
      cout << array[y][x] << " ";
    cout << endl;
  }
}


ostream& operator<<(ostream& out, const Matrix& obj){ //<<연산자 오버로딩
  out << "Matrix(" << obj.dy << "," << obj.dx << ")" << endl;
  for(int y = 0; y < obj.dy; y++){
    for(int x = 0; x < obj.dx; x++)
      out << obj.array[y][x] << " ";
    out << endl;
  }
  out << endl; //어차피 friend 함수니까 obj.print()쓰면 안되는건가..?
  return out;
}

Matrix& Matrix::operator=(const Matrix& obj) //=연산자 오버로딩
{
  if (this == &obj) return *this; 
  if ((dx != obj.dx) || (dy != obj.dy)) 
    alloc(obj.dy, obj.dx);

  for (int y = 0; y < dy; y++)
    for (int x = 0; x < dx; x++)
      array[y][x] = obj.array[y][x];
  return *this;
}

File: pytet_v0.5/matrix.py
class MatrixError(Exception):
    pass
#어떻게 보면 x,y가 행,열 일 수 있고 어떻게 보면 열,행 일 수 있음 (나는 열,행으로 볼 거임)
class Matrix:
    nAlloc = 0
    nFree = 0
    
    #Matrix의 할당된 ? 갯수
    def get_nAlloc(self):
        return Matrix.nAlloc
    
    #Matrix의 해제된 ? 갯수
    def get_nFree(self):
        return Matrix.nFree
    
    #블록의 y좌표(행좌표) 받기
    def get_dy(self):
        return self._dy
    
    #블록의 x좌표(열좌표) 받기
    def get_dx(self):
        return self._dx
    
    #블록의 배열 받기
    def get_array(self):
        return self._array
    
    #Matrix ? 해제
    def __del__(self):
        Matrix.nFree += 1
    
    #블록을 Matrix에 할당(붙여넣기) 
    def __alloc(self, cy, cx):
        if cy < 0 or cx < 0:
            raise MatrixError("wrong matrix size")
        self._dy = cy
        self._dx = cx
        self._array = [[0]*self._dx for i in range(self._dy)]
        #print(self.__array)
        Matrix.nAlloc += 1
        
    #Matrix 생성자      
    def __init__(self, arg):
        if isinstance(arg, list):
            array = arg
            cy = len(array)
            cx = len(array[0])
            self.__alloc(cy, cx)
            for y in range(cy):
                for x in range(cx):
                    self._array[y][x] = array[y][x]
            return
        elif isinstance(arg, Matrix):
            other = arg
            cy = other._dy
            cx = other._dx
            self.__alloc(cy, cx)
            for y in range(cy):
                for x in range(cx):
                    self._array[y][x] = other._array[y][x]
            return
        else:
            self.__alloc(0, 0)
            return
    
    def __str__(self):
        return 'Matrix(%d, %d)' % (self._dy, self._dx)

    #Matrix(self) 출력
    def print(self):
        print('[', end=' ')
        for y in range(self._dy-1):
            print('[', end=' ')
            for x in range(self._dx-1):
                print(self._array[y][x], end=', ')
            print(self._array[y][self._dx-1], end=' ')
            print('],', end=' ')
        print('[', end=' ')
        for x in range(self._dx-1):
            print(self._array[self._dy-1][x], end=', ')
        print(self._array[self._dy-1][self._dx-1], end=' ')
        print(']', end=' ')
        print(']')        
        return

    #Matrix(self)의 일정부분 잘라서 복사
    def clip(self, top, left, bottom, right):
        cy = bottom - top
        cx = right - left
        temp = [[0]*cx for i in range(cy)]       
        for y in range(cy):
            for x in range(cx):
                if (top+y >= 0) and (left+x >= 0) \
                   and (top+y < self._dy) and (left+x < self._dx):
                    temp[y][x] = self._array[top+y][left+x]
                else:
                    raise MatrixError("invalid matrix range")
        return Matrix(temp)
   
    #Matrix(self)에 other 붙여넣기
    def paste(self, other, top, left):
        for y in range(other._dy):
            for x in range(other._dx):
                if (top+y >= 0) and (left+x >= 0) \
                   and (top+y < self._dy) and (left+x < self._dx):
                    self._array[top+y][left+x] = other._array[y][x]
                else:
                    raise MatrixError("invalid matrix range")

    
    #Matrix 덧셈 (self+other)
    def __add__(self, other):
        if (self._dx != other._dx) or (self._dy != other._dy):
            raise MatrixError("matrix sizes mismatch")
        temp = [[0]*self._dx for i in range(self._dy)]
        for y in range(self._dy):
            for x in range(self._dx):
                temp[y][x] = self._array[y][x] + other._array[y][x]                
        return Matrix(temp)
    
    #Matrix 각 원소의 총 합
    def sum(self):
        total = 0
        for y in range(self._dy):
            for x in range(self._dx):
                total += self._array[y][x]
        return total
    
    #Matrix에 상수 곱하기 (self*coef)
    def mulc(self, coef):
        for y in range(self._dy):
            for x in range(self._dx):
                self._array[y][x] *= coef
    
    #Matrix의 각 원소가 특정 값보다 큰지 확인하기 (self>val=?)
    def anyGreaterThan(self, val):
        for y in range(self._dy):
            temp = [v for v in self._array[y] if v > val]
            if len(temp) > 0:
                return True
        return False
    
    #Matrix의 각 원소 값을 이진수로 나타내기 (0,1)
        temp = Matrix(self)
        for y in range(self._dy):
            for x in range(self._dx):
                if temp._array[y][x] != 0:
                    temp._array[y][x] = 1
        return temp

### end of Matrix class

